emcc.py

```diff
@@ -2239,23 +2254,22 @@ def check_memory_setting(setting):
     if settings.ALLOW_MEMORY_GROWTH:
       max_mem = settings.MAXIMUM_MEMORY
 
-    shadow_size = max_mem // 8
-    settings.GLOBAL_BASE = shadow_size
-
-    sanitizer_mem = (shadow_size + webassembly.WASM_PAGE_SIZE) & ~webassembly.WASM_PAGE_SIZE
-    # sanitizers do at least 9 page allocs of a single page during startup.
-    sanitizer_mem += webassembly.WASM_PAGE_SIZE * 9
-    # we also allocate at least 11 "regions". Each region is kRegionSize (2 << 20) but
-    # MmapAlignedOrDieOnFatalError adds another 2 << 20 for alignment.
-    sanitizer_mem += (1 << 21) * 11
-    # When running in the threaded mode asan needs to allocate an array of kMaxNumberOfThreads
-    # (1 << 22) pointers.  See compiler-rt/lib/asan/asan_thread.cpp.
-    if settings.USE_PTHREADS:
-      sanitizer_mem += (1 << 22) * 4
+    # The total memory is 1/8th shadow and 7/8th usable memory
+    mem_size_including_shadow = max_mem * 8 / 7
 
-    # Increase the size of the initial memory according to how much memory
-    # we think the sanitizers will use.
-    settings.INITIAL_MEMORY += sanitizer_mem
+    # But we might need to re-align to wasm page size
+    mem_size_including_shadow = int(round_to(mem_size_including_shadow, webassembly.WASM_PAGE_SIZE))
+
+    # The shadow size is 1/8th the resulting rounded up size
+    shadow_size = mem_size_including_shadow // 8
+
+    # We start our global data after the shadow memory
+    settings.GLOBAL_BASE = shadow_size + 1
```

Comment:
Does this not need to be aligned? Or does wasm-ld align it?

Reply:
Yes, wasm-ld will align the start of globals according to the alignment requirement of the first global.. I'll add a comment

Reply:
Is the `+ 1` necessary here? I would expect that the shadow region would be `[0, shadow_size)`, so starting static data at `shadow_size` would not overlap.

Reply:
Hmm yes, I think you are correct.  I forgot to do a `-1` when calculating the max address:

```
uptr max_address = (__builtin_wasm_memory_size(0) * WASM_PAGE_SIZE) - 1;
```